uswds/uswds
# USWDS - Range: Remove redundant ARIA attributes

# Summary

**Removed redundant ARIA attributes to improve the screen reader experience.** To incorporate these changes, update your range component markup.

## Breaking change

:warning: This is a breaking change.

## Related issue

Closes #5038 

## Related pull requests

[Changelog →](https://github.com/uswds/uswds-site/pull/2220)

## Preview link

Preview link:
[Range slider →](https://federalist-3b6ba08e-0df4-44c9-ac73-6fc193b0e19c.sites.pages.cloud.gov/preview/uswds/uswds/cm-range-slider-sr/iframe.html?args=&id=components-form-inputs-range--range)

## Problem statement

Range slider keyboard navigation does not work with **VoiceOver**. This is caused by redundant aria attributes that are reserved for elements that are more or less "faking" being an `input` with `type="range"`.

<details>

<summary>Example</summary>

Slider can only move one increment and does not update read out correctly:
_VoiceOver - Video has audio if you unmute_

https://github.com/uswds/uswds/assets/25211650/e4bed68f-fd87-408f-b489-6dc647d4bb62

</details>


**Note: Some more advanced screen readers such as JAWS may intuitively navigate around this issue**

## Solution

Remove the following attributes:
-  `aria-valuemin`
- `aria-valuemax`
- `aria-valuenow`
- `role="slider"`

### Sources

[MDN Docs: Range slider best practices →](https://developer.mozilla.org/en-US/docs/Web/Accessibility/ARIA/Roles/slider_role#best_practices)

<details>

<summary>Example</summary>

Slider has full range and updates read out correctly

https://github.com/uswds/uswds/assets/25211650/c7395539-04ed-425e-8542-895bbaf1b37a

</details>


## Testing and review

1. Test [range slider on develop branch →](https://federalist-3b6ba08e-0df4-44c9-ac73-6fc193b0e19c.sites.pages.cloud.gov/preview/uswds/uswds/develop/iframe.html?args=&id=components-form-inputs-range--range)
    1. Confirm keyboard navigation works without screen reader
    2. Activate the screen reader and attempt to use keyboard controls 
2. Open [range slider fix preview →](https://federalist-3b6ba08e-0df4-44c9-ac73-6fc193b0e19c.sites.pages.cloud.gov/preview/uswds/uswds/cm-range-slider-sr/iframe.html?args=&id=components-form-inputs-range--range)
3. Confirm screen reader keyboard navigation is possible
4. Confirmed read outs are improved
5. Confirm no visual or usability regression

### 
- [ ] Confirm screen reader navigation is possible
- [ ] Confirm screen reader readouts work as intended
- [ ] Confirm no regression in component

---

We have discussed potential screen reader callout enhancements which are captured in #5431
